node/src/types/node_config.rs

```diff
@@ -1,13 +1,22 @@
 use datasize::DataSize;
 use serde::{Deserialize, Serialize};
 
-use crate::types::BlockHash;
+use crate::types::{BlockHash, TimeDiff};
 
-/// Node configuration.
-#[derive(Default, DataSize, Debug, Deserialize, Serialize)]
+/// Node fast-sync configuration.
+#[derive(Default, DataSize, Debug, Deserialize, Serialize, Clone)]
 // Disallow unknown fields to ensure config files and command-line overrides contain valid keys.
 #[serde(deny_unknown_fields)]
 pub struct NodeConfig {
-    /// Hash used as a trust anchor when joining, if any.
+    /// Hash used as a trust anchor when joining. If no hash is provided, attempt to run genesis.
     pub trusted_hash: Option<BlockHash>,
+
+    /// An (optional) time to expiration for the header for the trusted hash. If the timestamp of
+    /// the trusted hash is less than the current time minus this expiration time, the node will not
+    /// join the network.
+    pub trusted_hash_time_to_expiration: Option<TimeDiff>,
```

Comment:
Can we compute that from the unbonding delay instead?

Reply:
You mentioned this above, and I am not sure how implement this.

Is the idea to set this to `era_length * unbonding_delay` rather than just a flat time diff?

Reply:
Or `minimum_era_length * (unbonding_delay - auction_delay)`… or, to be completely precise:

Let's try to find out whether a validator's signature in era `n` is still trustworthy. To do that, we'd have to find out the `start_time` of era `n`.
Since they were a validator in era `n` that means they can't have made an unbonding request before era `n - auction_delay`. If they made one in `n - auction_delay`, they will be able to withdraw their stake at the end of era `n - auction_delay + unbonding_delay`. (Is that correct?)
So the earliest time they could withdraw would be `start_time + minimum_era_length * (unbonding_delay + 1 - auction_delay)`.

That definitely needs to be double-checked to make sure I didn't make any off-by-one errors, and that I correctly understand how the unbonding delay works!

But I still think it would be safer to compute the trusted time this way, or to use some lower bound like `minimum_era_length * (unbonding_delay - auction_delay)`.

Reply:
Technically, this doesn't have to be part of this PR - this one is *only* about archival nodes. We can solve that problem later in a different PR.

Reply:
Sure, it's not a blocker; it's just that we'd introduce a configuration option that we'd end up removing/deprecating again.

Reply:
I meant to remove that change from this PR entirely so to not block otherwise correct PR.

Reply:
This was removed in https://github.com/casper-network/casper-node/pull/1995/commits/fe6cd1be96a832d660077068a208a5eaa5f1135e